src/core/raster/qgshillshaderenderer.cpp

```diff
@@ -0,0 +1,249 @@
+  double sinzenith_rad = cos( zenith_rad );
+
+  QRgb myDefaultColor = NODATA_COLOR;
+
+  for ( qgssize i = 0; i < ( qgssize )height; i++ )
+  {
+
+    for ( qgssize j = 0; j < ( qgssize )width; j++ )
+    {
+
+      if ( inputBlock->isNoData( i, j ) )
+      {
+        outputBlock->setColor( i, j, myDefaultColor );
+        continue;
+      }
+
+      if ( inputBlock->isNoData( i, j ) )
+      {
+        outputBlock->setColor( i, j, myDefaultColor );
+        continue;
+      }
+
+      qgssize iUp, iDown, jLeft, jRight;
+      if ( i == 0 )
+      {
+        iUp = i;
+        iDown = i + 1;
+      }
+      else if ( i < ( qgssize )height - 1 )
+      {
+        iUp = i - 1;
+        iDown = i + 1;
+      }
+      else
+      {
+        iUp = i - 1;
+        iDown = i;
+      }
+
+      if ( j == 0 )
+      {
+        jLeft = j;
+        jRight = j + 1;
+      }
+      else if ( j < ( qgssize )width - 1 )
+      {
+        jLeft = j - 1;
+        jRight = j + 1;
+      }
+      else
+      {
+        jLeft = j - 1;
+        jRight = j;
+      }
+
+      double x11;
+      double x21;
+      double x31;
+      double x12;
+      double x22; // Working cell
+      double x32;
+      double x13;
+      double x23;
+      double x33;
+
+      // This is center cell. It is not nodata. Use this in place of nodata neighbors
+      x22 = inputBlock->value( i, j );
+
+      x11 = inputBlock->isNoData( iUp, jLeft )  ? x22 : inputBlock->value( iUp, jLeft );
+      x21 = inputBlock->isNoData( i, jLeft )     ? x22 : inputBlock->value( i, jLeft );
+      x31 = inputBlock->isNoData( iDown, jLeft ) ? x22 : inputBlock->value( iDown, jLeft );
+
+      x12 = inputBlock->isNoData( iUp, j )       ? x22 : inputBlock->value( iUp, j );
+      // x22
+      x32 = inputBlock->isNoData( iDown, j )     ? x22 : inputBlock->value( iDown, j );
+
+      x13 = inputBlock->isNoData( iUp, jRight )   ? x22 : inputBlock->value( iUp, jRight );
+      x23 = inputBlock->isNoData( i, jRight )     ? x22 : inputBlock->value( i, jRight );
+      x33 = inputBlock->isNoData( iDown, jRight ) ? x22 : inputBlock->value( iDown, jRight );
+
+      double derX = calcFirstDerX( x11, x21, x31, x12, x22, x32, x13, x23, x33, cellxsize );
+      double derY = calcFirstDerY( x11, x21, x31, x12, x22, x32, x13, x23, x33, cellysize );
+
+      double slope_rad = atan( mZFactor * sqrt( derX * derX + derY * derY ) );
+
+      if ( derX != 0 )
+      {
+        aspect_rad = atan2( derX, -derY );
+        if ( aspect_rad < 0 )
+        {
+          aspect_rad = 2 * M_PI + aspect_rad;
+        }
+      }
+      else if ( derX == 0 )
+      {
+        if ( derY > 0 )
+        {
+          aspect_rad = M_PI / 2;
+        }
+        else if ( derY < 0 )
+        {
+          aspect_rad = 2 * M_PI - M_PI / 2;
+        }
+        else
+        {
+          aspect_rad = aspect_rad;
+        }
+      }
+
+
+      double colorvalue = qMax( 0.0, 255.0 * (( coszenith_rad * cos( slope_rad ) ) +
```

Comment:
Go:

```
      double colorvalue = qBound( 0.0, 255.0 * (( coszenith_rad * cos( slope_rad ) ) +
                                              ( sinzenith_rad * sin( slope_rad ) * cos( azimuth_rad - aspect_rad ) ) ), 255.0 );
```
instead, so that colors can be at most 255. Fixes the black pixels I see on high altitudes.